src/PhpDoc/TypeNodeResolver.php

```diff
@@ -702,11 +702,11 @@ private function resolveConstTypeNode(ConstTypeNode $typeNode, NameScope $nameSc
 			$classReflection = $this->getReflectionProvider()->getClass($className);
 
 			$constantName = $constExpr->name;
-			if (Strings::endsWith($constantName, '*')) {
-				$constantNameStartsWith = Strings::substring($constantName, 0, Strings::length($constantName) - 1);
+			if (Strings::contains($constantName, '*')) {
+				$pattern = '{^' . str_replace('\\*', '.*?', preg_quote($constantName)) . '$}D';
```

Comment:
Calling https://www.php.net/manual/en/function.fnmatch.php might be an easier way to do this?

Reply:
Not so sure.. Because that also matches `[]` blocks etc, might do more than we want it to, although I guess the lexer won't ever give you those characters in the constant name. This preg way is easy and works, but if you really want I can do it with fnmatch.

Reply:
Yeah, this is fine if it works, but I have to admit I'm not a regex expert so I have no idea what's going on in the current code :)

Reply:
Added a comment if it helps ;) And more tests to check multiple `*`s. 

Just to explain in more details what this does:

- preg_quote escapes *all* regex metacharacters, so matching $candidate against `'{^' . preg_quote($constantName) . '$}D'` would simply be equivalent to doing `$constantName === $candidate`
- That's not so helpful of course so adding the str_replace turns back `\*` (i.e. escaped wildcards into `.*?` so you end up with a pattern like `FOO*` => `FOO.*?` which allows for anything where the wildcard was - The complete pattern is `{^FOO.*?$}D` though to make sure it does not match a part of the string, the anchors ^/$ make sure it's matching the whole candidate name.
- The `D` modifier ensures that `$` cannot match if there is a newline, without it `^FOO.*?$` would match also `FOOBAR\n` which we don't want even though here I guess it is not possible to get such a string to be matched against anyway.

Reply:
Awesome, thank you :)